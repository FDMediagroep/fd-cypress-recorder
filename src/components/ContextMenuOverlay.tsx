import { PureComponent } from "react";
import React from "react";
import styled from "styled-components";
import ContextMenu from "./ContextMenu";

export interface Props {
    target: HTMLElement;
    selector: string;
    onClick: (e: React.MouseEvent<HTMLElement>) => void;
}

/**
 * This component renders the overlay for the context menu and also contains the context menu.
 */
export default class ContextMenuOverlay extends PureComponent<Props, any> {
    render() {
        return (
            <StyledContextOverlay onClick={this.props.onClick}>
                <StyledContextContainer
                    top={document.documentElement.scrollTop + this.props.target.getBoundingClientRect().top}
                    width={this.props.target.getBoundingClientRect().width}
                    height={this.props.target.getBoundingClientRect().height}
                    left={document.documentElement.scrollLeft + this.props.target.getBoundingClientRect().left}
                />
                <ContextMenu target={this.props.target} selector={this.props.selector}/>
            </StyledContextOverlay>
        );
    }
}

const StyledContextContainer: any = styled.div`
    position: absolute;
    box-shadow: 0 0 0 1000px rgba(0,0,0,.5);
    /* for real browsers */
    box-shadow: 0 0 0 100vmax rgba(0,0,0,.5);
    top: ${(props: any) => props.top}px;
    width: ${(props: any) => props.width}px;
    height: ${(props: any) => props.height}px;
    left: ${(props: any) => props.left}px;
`;

const StyledContextOverlay = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 9999;
`;
