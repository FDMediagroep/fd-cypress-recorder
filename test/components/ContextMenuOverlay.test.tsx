import React from 'react';
import {act} from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import ContextMenuOverlay from '../../src/components/ContextMenuOverlay';

describe('Context Menu Overlay', () => {
    let target: HTMLElement;
    const clickEvt = new MouseEvent('click', {bubbles: true});
    const clickMock = jest.fn();

    beforeEach(() => {
        target = document.createElement('div');
        target.setAttribute('class', 'context-menu-overlay-test');
        document.body.appendChild(target);
        act(() => {
            ReactDOM.render(<ContextMenuOverlay onClick={clickMock} target={target} selector={'.context-menu-overlay-test'}/>, target);
        });
    });

    afterEach(() => {
        target.removeEventListener('click', clickMock);
        document.body.removeChild(target);
        target.remove();
    });

    it('should render Context Menu Overlay correctly', () => {
        let contextMenu = TestRenderer.create(<ContextMenuOverlay onClick={clickMock} target={target} selector={'.context-menu-overlay-test'}/>);
        expect(contextMenu.toJSON()).toMatchSnapshot();
    });

    it('should handle Click event correctly', () => {
        const contextMenuOverlay = target.querySelector('div');
        if (contextMenuOverlay) {
            contextMenuOverlay.dispatchEvent(clickEvt);
        }
        expect(clickMock).toHaveBeenCalledTimes(1);
    });

});
