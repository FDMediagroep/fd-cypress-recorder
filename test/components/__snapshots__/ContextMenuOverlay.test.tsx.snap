// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Menu Overlay should render Context Menu Overlay correctly 1`] = `
<div
  className="sc-ifAKCX EmPKX"
  onClick={[MockFunction]}
>
  <div
    className="sc-EHOje hFkCuy"
    onClick={[MockFunction]}
  />
  <div
    className="sc-bxivhb dzPEtI"
    height={0}
    width={0}
  />
  <div
    className="sc-htpNat eiGWKe"
  >
    <ul>
      <li
        className="label"
      >
        Interactions
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Click
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Enter text...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Hover
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Wait...
      </li>
      <li
        className="label"
      >
        Asserts
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Attributes...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Contains text...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Count...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Exists
      </li>
      <li
        className="label"
      >
        Global
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Go to URL...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Match current URL
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        URL contains...
      </li>
      <li
        className="clickable"
        onMouseDown={[Function]}
      >
        Visit current URL
      </li>
    </ul>
    <small>
      .context-menu-overlay-test
    </small>
  </div>
</div>
`;
